### PR TITLE
Implement `NUMERIC` overload for arithmetic functions

### PR DESCRIPTION
**Pull request by matriv**
- ABS
- CEIL
- FLOOR
- SQRT
- EXP
- ROUND (without precision)

Closes: #15632

### CODE REVIEW

File: docs/appendices/release-notes/5.10.0.rst [hunk 1]
```diff
@@ -69,7 +69,10 @@ None
 Data Types
 ----------
 
-None
+- Added support for :ref:`NUMERIC type<type-numeric>` to the following
+  arithmetic scalar functions: :ref:`ABS<scalar-abs>`, :ref:`CEIL<scalar-ceil>`,
+  :ref:`FLOOR<scalar-floor>`, :ref:`ROUND<scalar-round>`, :ref:`EXP<scalar-exp>`
+  and :ref:`SQRT<scalar-sqrt>`.
```
Comment (BaurzhanSakhariev) [inline docs/appendices/release-notes/5.10.0.rst:75]: afaik one of the ROUND versions was working before with NUMERIC. 
Can we specify more precisely which version of ROUND got new signature?

File: server/src/main/java/io/crate/expression/scalar/arithmetic/AbsFunction.java [hunk 2]
```diff
@@ -53,5 +57,21 @@ public static void register(Functions.Builder builder) {
...
             );
         }
+        builder.add(
+            Signature.builder(NAME, FunctionType.SCALAR)
+                .argumentTypes(DataTypes.NUMERIC.getTypeSignature())
+                .returnType(DataTypes.NUMERIC.getTypeSignature())
+                .features(Scalar.Feature.DETERMINISTIC, Scalar.Feature.STRICTNULL)
+                .build(),
+            (signature, boundSignature) -> {
+                DataType<?> argType = boundSignature.argTypes().get(0);
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/expression/scalar/arithmetic/AbsFunction.java:67]: I guess getting boundSignature argType and using it as a return type is required to preserve arg params?
Can we add a short code comment here and in there places ? Similar to how it's done in `NumericSumAggregation`.

**Review by BaurzhanSakhariev — DISMISSED** [review]
Added some minor comments, LGTM otherwise

File: server/src/main/java/io/crate/expression/scalar/arithmetic/AbsFunction.java (hunk 2 above)
Comment (matriv) [reply server/src/main/java/io/crate/expression/scalar/arithmetic/AbsFunction.java:67]: yep, thx!

File: server/src/test/java/io/crate/expression/scalar/arithmetic/SquareRootFunctionTest.java [hunk 3]
```diff
@@ -39,12 +41,15 @@ public void testSqrt() {
         assertEvaluateNull("sqrt(null)");
         assertEvaluate("sqrt(cast(25 as integer))", 5.0);
         assertEvaluate("sqrt(cast(25.0 as float))", 5.0);
+        assertEvaluate("sqrt(cast(225.0 as numeric(6, 1)))", new BigDecimal("15.0"));
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/expression/scalar/arithmetic/SquareRootFunctionTest.java:44]: can we add an example where answer is not integer, so demostrate that argument precision/scale is applied?

Comment (matriv) [reply server/src/test/java/io/crate/expression/scalar/arithmetic/SquareRootFunctionTest.java]: done

File: server/src/test/java/io/crate/expression/scalar/arithmetic/SquareRootFunctionTest.java [hunk 4]
```diff
@@ -39,12 +41,16 @@ public void testSqrt() {
         assertEvaluateNull("sqrt(null)");
         assertEvaluate("sqrt(cast(25 as integer))", 5.0);
         assertEvaluate("sqrt(cast(25.0 as float))", 5.0);
+        assertEvaluate("sqrt(cast(123.4 as numeric(6, 1)))",
+            new BigDecimal("11.10855526159905278255972911272118"));
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/expression/scalar/arithmetic/SquareRootFunctionTest.java:45]: Shouldn't it have less digits?

Comment (matriv) [reply server/src/test/java/io/crate/expression/scalar/arithmetic/SquareRootFunctionTest.java:45]: postgres doesn't seem to respect the input arg, and truncate accordingly the result:
```
matriv=> select sqrt(cast(123.4 as numeric(4, 1)));
        sqrt        
--------------------
 11.108555261599053
(1 row)
```
and imho we should not preserve the type for `sqrt` and `exp`. It's kind of the same behavior that those 2 functions when they take an `int` arg, the return `double`, the highest available precision type we have. So for numeric of any precision/scale we should probably return just `DataTypes.NUMERIC` with null precision and scale.

Comment (jeeminso) [reply server/src/test/java/io/crate/expression/scalar/arithmetic/SquareRootFunctionTest.java:45]: > On the contrary, I believe, abs/ceil/floor/round should preserve it.

Could we also consider not preserving precision/scale for Postgres compatibility? Or for consistency across numeric operations.

Comment (matriv) [reply server/src/test/java/io/crate/expression/scalar/arithmetic/SquareRootFunctionTest.java:45]: We could also do that, yes.

Comment (matriv) [reply server/src/test/java/io/crate/expression/scalar/arithmetic/SquareRootFunctionTest.java:45]: As discussed, we chose, for the time being, to return generic `NUMERIC` without precision/scale, for all those functions.

**Review by BaurzhanSakhariev — APPROVED** [review]
Thanks!

I was wondering if we should emphasize that behavior in docs - but I think 
> Return value will be of type ``numeric`` if the input value is of ``numeric``

is good enough. It's other way around: if we ever decide to enforce arg precision, than we should mention it in docs.

File: server/src/test/java/io/crate/expression/scalar/arithmetic/SquareRootFunctionTest.java (hunk 4 above)
Comment (matriv) [reply server/src/test/java/io/crate/expression/scalar/arithmetic/SquareRootFunctionTest.java:45]: Added comment to relevant issue, to revisit this behavior consistently: https://github.com/crate/crate/issues/16478#issuecomment-2426200103

**Comment by matriv** [comment]
> Thanks!
> 
> I was wondering if we should emphase that behavior in docs - but I think
> 
> > Return value will be of type `numeric` if the input value is of `numeric`
> 
> is good enough. It's other way around: if we ever decide to enforce arg precision, than we should mention it in docs.

Thx, I thought of it as well, but I also agree to not add something else there, until we decide if we leave it as is or we change the behavior globally.

**Pull request merged**